added tracking song updates with kafka.

Aliaksandr-Litvinau opened the pull request:
@evt
in an attempt to find the perfect solution in a vacuum)

Commit:
added tracking song updates with kafka.

evt commented on internal/app/kafka/services/runner/runner.go:
```diff
@@ -0,0 +1,54 @@
+
+	// Creating song updates service with Kafka
+	updateService, err := song_updates.NewSongUpdateService(
+		[]string{"kafka:9092"},
```
must be a config setting

Aliaksandr-Litvinau replied:
done. + added sync once for config.

evt commented on internal/app/kafka/services/song_updates/service.go:
```diff
@@ -0,0 +1,175 @@
+
+// SongUpdateService controls the processing of song updates
+type SongUpdateService struct {
+	producer kafka.Producer
```
It's usually a good idea to keep interface in the same package where you use it. Having interface in a separate package makes sense when you have a lot of similar interfaces used in different places, but again it's something to discuss with your team.

Aliaksandr-Litvinau replied:
It seemed to me that separating interfaces into a separate **package kafka/interfaces**, which is used by both implementations **(producer.go, consumer.go)** and services **(song_updates/service.go)**, would be a **compromise architectural solution**. This does not violate the principle of "interface where it is used", but on the contrary, creates a clear contract between the system components.
Such code organization follows the dependency inversion principle (DIP) from SOLID: high-level modules (services) and low-level modules (Kafka implementations) depend on abstractions (interfaces), and not on each other. This avoids tight coupling between components and makes the system more flexible and maintainable.
Common interfaces in a separate package also simplify testing (easy to create mocks), allow having multiple implementations of one interface (for example, for different message brokers or test environments), and provide a single contract for all components working with Kafka. At the same time, each component can be sure that any implementation of these interfaces will correspond to the expected behavior. As a result, such a structure makes the code more modular, testable and resistant to changes, which is fully consistent with the principles of clean architecture and modern development practices.

evt commented on internal/app/kafka/services/song_updates/service.go:
```diff
@@ -0,0 +1,175 @@
+		defer s.wg.Done()
+		if err := s.consumer.Start(ctx); err != nil {
+			log.Printf("Error starting consumer: %v", err)
+			s.errChan <- fmt.Errorf("consumer error: %w", err)
```
sometimes it's a good idea to use this package to handle errors from goroutines: https://pkg.go.dev/golang.org/x/sync/errgroup

Aliaksandr-Litvinau replied:
done

evt commented on internal/app/kafka/services/song_updates/service.go:
```diff
@@ -0,0 +1,175 @@
+						cancel()
+						return
+					}
+				} else {
```
better use `continue` than `else`

evt commented on internal/app/kafka/services/song_updates/service.go:
```diff
@@ -0,0 +1,175 @@
+		s.cancel()
+	}
+
+	// Wait for all goroutines to complete
+	done := make(chan struct{})
+	go func() {
+		s.wg.Wait()
+		close(done)
+	}()
```
I don't quite get why you need this code here.

Aliaksandr-Litvinau replied:
rewrite 
`go func() {
    defer close(s.done)
    if err := g.Wait(); err != nil {
        log.Printf("Service error occurred: %v", err)
    }
    cancel() 
}()`

Commit:
fixed config Kafka with sync once.

Commit:
error handling fix for kafka.

Commit:
Merge branch 'refs/heads/master' into feature/kafka

# Conflicts:
#	cmd/main.go
#	go.mod
#	go.sum
#	internal/app/config/config.go

Commit:
added graceful shutdown.

Aliaksandr-Litvinau commented:
@evt 
Additional changes after CR:
**1. Parallel service startup**
HTTP, gRPC, and Kafka now start simultaneously, which speeds up application startup.
**2. Sequential shutdown**
Services shut down in strict order: first HTTP and gRPC, then Kafka, and finally the database connection. This ensures that operations are completed correctly.
**3. Waiting for Kafka readiness**
Added a smart wait for a connection to Kafka for a minute to avoid errors during fast container startup.
**4. Individual shutdown timeouts**
Each component has its own shutdown timeouts. If one of the services hangs, it does not block the shutdown of the others.
5. Replaced outdated balancing strategies (BalanceStrategyRoundRobin, BalanceStrategySticky) with the modern BalanceStrategyRangeV2, which works better with reconnections and partition rebalancing.

Commit:
updated Kafka interfaces.

Commit:
updated error's handling.

Aliaksandr-Litvinau committed:
Added DB wrapper with connection pooling, logging and health checks following Open/Closed principle (SOLID) to allow extension without modification.

Aliaksandr-Litvinau committed:
fixed sync once in config, updated compose file.

Aliaksandr-Litvinau committed:
added DB connection close handling via defer.

Aliaksandr-Litvinau committed:
delete *gorm.DB embedding.

Aliaksandr-Litvinau committed:
Removed id parameter from UpdateSong method, added a check for nil.

Aliaksandr-Litvinau committed:
Added DateFormat constant for consistent YYYY-MM-DD formatting, replace hardcoded "2006-01-02", fix API response to use shorter date format without time.

Aliaksandr-Litvinau committed:
fixed after goimports -l -w .

Aliaksandr-Litvinau committed:
updated docker-compose file for local development:
- added Kafka UI service,
- configure ports for
internal/external communication,
- set up environment variables for local usage.

Aliaksandr-Litvinau committed:
added using external mock API for enrich song.

Aliaksandr-Litvinau committed:
fixed tests.
